### ISSUE TITLE
Uvicorn >=0.11.4 does not send shutdown event when using --reload

### ISSUE THREAD

**Issue by ghost** [Kludex/uvicorn#679]
Hello, I have a problem. After update Uvicorn from 0.11.3 to 0.11.5, it does not send shutdown event.

Source code:
```python
from starlette.applications import Starlette

def startup():
    print('Ready to go')

def shutdown():
    print('Shutdown!')

app = Starlette(debug=True, on_startup=[startup], on_shutdown=[shutdown])
```

Uvicorn 0.11.3:
```
[uvicorn-ERV875EY]─[~/p/uvicorn]─> uvicorn --version
Running uvicorn 0.11.3 with CPython 3.8.2 on Darwin

[uvicorn-ERV875EY]─[~/p/uvicorn]─> uvicorn --reload --log-level trace main:app
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [47131]
INFO:     Started server process [47134]
INFO:     Waiting for application startup.
TRACE:    ASGI [1] Started scope={'type': 'lifespan'}
TRACE:    ASGI [1] Receive {'type': 'lifespan.startup'}
Ready to go
TRACE:    ASGI [1] Send {'type': 'lifespan.startup.complete'}
INFO:     Application startup complete.
^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
TRACE:    ASGI [1] Receive {'type': 'lifespan.shutdown'}
Shutdown!
TRACE:    ASGI [1] Send {'type': 'lifespan.shutdown.complete'}
TRACE:    ASGI [1] Completed
INFO:     Application shutdown complete.
INFO:     Finished server process [47134]
INFO:     Stopping reloader process [47131]
```

Uvicorn 0.11.4:
```
[uvicorn-ERV875EY]─[~/p/uvicorn]─> uvicorn --version
Running uvicorn 0.11.4 with CPython 3.8.2 on Darwin

[uvicorn-ERV875EY]─[~/p/uvicorn]─> uvicorn --reload --log-level trace main:app
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [47903] using statreload
INFO:     Started server process [47905]
INFO:     Waiting for application startup.
TRACE:    ASGI [1] Started scope={'type': 'lifespan'}
TRACE:    ASGI [1] Receive {'type': 'lifespan.startup'}
Ready to go
TRACE:    ASGI [1] Send {'type': 'lifespan.startup.complete'}
INFO:     Application startup complete.
^CINFO:     Shutting down
INFO:     Finished server process [47905]
INFO:     Stopping reloader process [47903]
```

**Comment by florimondmanca** [comment]
Looks like this is only the case when `--reload` is passed.

Another data point against this pure-ASGI (no Starlette) app…

```python
async def app(scope, receive, send):
    assert scope["type"] == "lifespan"
    await lifespan(scope, receive, send)


async def lifespan(scope, receive, send):
    message = await receive()
    assert message["type"] == "lifespan.startup"
    print("Ready to go!")
    await send({"type": "lifespan.startup.complete"})

    message = await receive()
    assert message["type"] == "lifespan.shutdown"
    print("Shutdown!")
    await send({"type": "lifespan.shutdown.complete"})
```

With `--reload` - shutdown not performed:

```console
$ uvicorn debug.app:app --reload --log-level trace
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [25154] using statreload
INFO:     Started server process [25156]
INFO:     Waiting for application startup.
TRACE:    ASGI [1] Started scope={'type': 'lifespan'}
TRACE:    ASGI [1] Receive {'type': 'lifespan.startup'}
Ready to go!
TRACE:    ASGI [1] Send {'type': 'lifespan.startup.complete'}
INFO:     Application startup complete.
^CINFO:     Shutting down
INFO:     Finished server process [25156]
INFO:     Stopping reloader process [25154]
```

Without `--reload` - shutdown is performed:

```console
$ uvicorn debug.app:app --log-level trace 
INFO:     Started server process [25394]
INFO:     Waiting for application startup.
TRACE:    ASGI [1] Started scope={'type': 'lifespan'}
TRACE:    ASGI [1] Receive {'type': 'lifespan.startup'}
Ready to go!
TRACE:    ASGI [1] Send {'type': 'lifespan.startup.complete'}
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
TRACE:    ASGI [1] Receive {'type': 'lifespan.shutdown'}
Shutdown!
TRACE:    ASGI [1] Send {'type': 'lifespan.shutdown.complete'}
TRACE:    ASGI [1] Completed
INFO:     Application shutdown complete.
INFO:     Finished server process [25394]
```

**Comment by florimondmanca** [comment]
Okay, so this is a regression that was introduced by #620.

When using `--reload`, signal handlers are installed:

- On 0.11.3, the reloader signal handler just sets `should_exit`, so that the reloader stops. Then `.shutdown()` is called, which sends the `SIGINT` signal into the server, which runs the shutdown logic.
- On 0.11.4, the reloader signal handler _also_ shuts down child server processes manually by sending them the `SIGINT` signal. Once this is done, `self.process.join()` is still called in the reloader's `.shutdown()`, which means that the process gets _another `SIGINT` signal. This is equivalent to pressing `Ctrl+C` twice, which force-exits the server, skipping any shutdown logic.

Commenting these lines brings shutdown back by only sending `SIGINT` once:

https://github.com/encode/uvicorn/blob/7e55554a9f32a127ea3511fa2f55d1b47b66029d/uvicorn/supervisors/basereload.py#L32-L37

Well, this would essentially revert #620, which will probably bring back the issue it was trying to fix, i.e. #364.

@euri10 Any thoughts? I had not followed discussions on #364 but it looks like there #620 might not have been exactly the right fix - see notes above. :-)

**Comment by euri10** [comment]
yes you're correct we should revert #620 as it does not handle the original issue properly
I thought I did it already to be fair, this one skipped in my inbox sorry

on a side note:

overall I've come to the conclusion that our shutdown process when using the reloader or multiprocess is broken, after we revert #620 (or before 0.11.3) you'll see that sending a simple SIGTERM doesn't gracefully shutdown the server, it hangs forever after the join so there's something going on.
I think that went unnoticed largely because the SIGINT case works fine, and it came up as a "docker" issue in #364 because under the hood docker send a SIGTERM. but in fact it's not in docker that the issue lies, it's uvicorn that doesn't shutdown gracefully in that scenario (sigterm + reloder / multiprocess).

If you look at how hypercorn handles the thing it's rather interesting, I have a poc, largely unfinished and unpolished that takes inspiration from it, and it solves what appears to me as a SIGTERM issue only, but that's another story

**Comment by euri10** [comment]
this one should be fixed in 0.12
